### ISSUE TITLE
planet troops regen even during combat

### ISSUE THREAD

**Issue by Dilvish-fo** [freeorion/freeorion#2015]
[As noted on the forums](http://www.freeorion.org/forum/viewtopic.php?f=28&t=10878), for over a week now planetary troops have been getting regen even when the planet is under attack.  I thought we had likely sorted out the issue with some corrections to the scripts, but that does not appear to be the case.  I have seen various levels of regen (and also with planets using Defense Focus or not); it does appear to coincide with the "regular" non-combat regen level.  The various Garrison tech scripts themselves all are gated by 
```
(LocalCandidate.System.LastTurnBattleHere < CurrentTurn)
```

Which does appear correct.  Just to check I tried swapping in ```LastTurnAttackedByShip``` in place of  ```LastTurnBattleHere``` but it made no difference.

**Comment by geoffthemedio** [comment]
Can you provide a save demonstrating the issue?

**Comment by Dilvish-fo** [comment]
Happening both at Nidor and at Gienah Beta.
[wrong_troop_regen.zip](https://github.com/freeorion/freeorion/files/1742067/wrong_troop_regen.zip)

**Comment by Dilvish-fo** [comment]
minor update: using Debug Data Dumps, I verified that LastTurnAttackedByShip (for the planet) and LastTurnBattleHere (for the system) are being updated properly.

**Comment by geoffthemedio** [comment]
That should be available in the objects list columns as well...

https://github.com/freeorion/freeorion/blob/master/UI/ObjectListWnd.cpp#L190

**Comment by geoffthemedio** [comment]
Possibly `LocalCandidate` isn't being populated as expected when evaluating conditions that include `(LocalCandidate.System.LastTurnBattleHere < CurrentTurn)`. Then when it tries to evaluate the left part of that, it returns -65535, or the "invalid game turn" value.

**Comment by Dilvish-fo** [comment]
>  Then when it tries to evaluate the left part of that, it returns -65535, or the "invalid game turn" value.

I had already done a test on that yesterday, with augmented logging for the path where that returns the invalid game turn value, and didn't get any hits from it...

**Comment by Dilvish-fo** [comment]
hmm, although it was also yesterday that I discovered that sometimes, not always, when I rebuild FO my logging options return to the default "Debug" on everything, I don't recall the relative timing of that discovery with my check for the invalid turn, so I'll repeat that check.

**Comment by geoffthemedio** [comment]
Looking at it more, I'm more inclined to think that's not the issue... Possibly it's returning -65535 (when it shouldn't) in some other cases, though...

**Comment by dbenage-cx** [comment]
Noticed at least the planet last turn attacked does not seem correct, which at least precludes availability of LastTurnAttackedByShip.  Possibly the system last battle is not correct at this point either (system being sluggish recompiling atm).

From planet being attacked for >3 turns:
```
[trace] effects : Effect.cpp:433 : 

Execute SetMeter effect: 
SetTroops value = min(Value + 1.00, Target.MaxTroops))

[trace] effects : Effect.cpp:434 : SetMeter execute targets before: 
[trace] effects : Effect.cpp:436 :  ... OBJ_PLANET 2398: Sarin α II  at: Sarin α owner: Syndicate created on turn: -32768 specials:   Meters: METER_TARGET_POPULATION: Cur: 14 Init: 14  METER_TARGET_INDUSTRY: Cur: 5 Init: 5  METER_TARGET_RESEARCH: Cur: 20.2 Init: 20.2  METER_TARGET_TRADE: Cur: 0 Init: 0  METER_TARGET_CONSTRUCTION: Cur: 30 Init: 30  METER_TARGET_HAPPINESS: Cur: 25 Init: 25  METER_MAX_SHIELD: Cur: 191 Init: 41  METER_MAX_DEFENSE: Cur: 45 Init: 45  METER_MAX_SUPPLY: Cur: 4 Init: 4  METER_MAX_STOCKPILE: Cur: 0.28 Init: 0.28  METER_MAX_TROOPS: Cur: 30.2 Init: 30.2  METER_POPULATION: Cur: 14 Init: 14  METER_INDUSTRY: Cur: 5 Init: 5  METER_RESEARCH: Cur: 20.2 Init: 20.2  METER_TRADE: Cur: 0 Init: 0  METER_CONSTRUCTION: Cur: 5 Init: 2  METER_HAPPINESS: Cur: 25 Init: 25  METER_SHIELD: Cur: 5 Init: 0  METER_DEFENSE: Cur: 19 Init: 9  METER_SUPPLY: Cur: 4 Init: 4  METER_STOCKPILE: Cur: 0.28 Init: 0.28  METER_TROOPS: Cur: 19.2 Init: 15.2  METER_REBEL_TROOPS: Cur: 0 Init: 0  METER_STEALTH: Cur: 5 Init: 5  METER_DETECTION: Cur: 150 Init: 150   species: SP_CHATO  ResourceCenter focus: FOCUS_RESEARCH last changed on turn: 143 type: PT_SWAMP original type: PT_SWAMP size: SZ_SMALL rot period: 0.824265 axis tilt: 13.6532 buildings:  conqured on turn: -65535 last attacked on turn: -1
```

**Comment by Dilvish-fo** [comment]
> Noticed at least the planet last turn attacked does not seem correct...

I just double checked my test-case again, via the Objects window, and both the planet LastTurnAttackedByShip and LastTurnBattleHere and the system LastTurnBattleHere are reading out correctly.  Plus, the below augmented logging (which I just now got) shows that while Gienah β I (the planet I have blockaded that is still regenning troops) does properly qualify for the DEF_GARRISON_I 
effect which has no LastTurnBattleHere check , but is properly NOT qualifying for the DEF_GARRISON_II regen effect which does have the LastTurnBattleHere check.  

```
15:10:07.827654 [trace] effects : Universe.cpp:1152 : StoreTargetsAndCausesOfEffectsGroups: effects_group:   specific_cause: DEF_GARRISON_1  sources: Glorious Sceptrum I (9086)  )
Generated new target set, for Condition:  that is a Planet and that belongs to source's owner empire    targets: (Minkar γ I (868)  Minkar γ II (882)  Minkar γ Asteroids (896)  Gienah β I (3150)  Bacabs α III (6272)  Bacabs α II (6286)  Bacabs α V (6300)  Bacabs α I (6314)  Bacabs α IV (6328)  Phaet α I (6384)  Aten β I (7196)  Glorious Sceptrum I (9086)  Marfik α I (7980)  Tyr I (8358)  )
```
```
15:10:07.908746 [trace] effects : Universe.cpp:1152 : StoreTargetsAndCausesOfEffectsGroups: effects_group:   specific_cause: DEF_GARRISON_2  sources: Glorious Sceptrum I (9086)  )
Generated new target set, for Condition:  that is a Planet and that belongs to source's owner empire and when local condition candidate's system's last turn a battle occurred here < the current turn    targets: (Minkar γ I (868)  Minkar γ II (882)  Minkar γ Asteroids (896)  Tyr I (8358)  Bacabs α III (6272)  Bacabs α II (6286)  Bacabs α V (6300)  Bacabs α I (6314)  Bacabs α IV (6328)  Phaet α I (6384)  Aten β I (7196)  Glorious Sceptrum I (9086)  Marfik α I (7980)  )
```
So it appears to me that the source of the trouble is neither the Garrison techs in general, nor the LastTurnBattleHere check in particular.  There appears to be some other source of regen, I am still trying to track that down.
 I think the augmented logging I show above is pretty helpful for this kind of thing, I'll put up a PR for it.

**Comment by Dilvish-fo** [comment]
I augmented the logging some more, in the spot that dbenage-cx was looking at, and that now clearly identifies that the troops were being increased by DEF_ROOT_DEFENSE, DEF_GARRISON_2, DEF_GARRISON_3 and SP_UGMORS (contrary to the other logging).  

```
 * * * * * * * * * * * (new effects group log entry)(DEF_ROOT_DEFENSE )
...
Execute SetMeter effect: 
SetShield value = min(Value + 1.00, Target.MaxShield))
... SetMeter execute targets before: 
17:08:47.961277 [trace] effects : Effect.cpp:437 :  ... OBJ_PLANET 3150: Gienah β I  ... METER_SHIELD: Cur: 0 Init: 0  ...  METER_TROOPS: Cur: 39.9 Init: 35.9  ... last attacked on turn: 117
... SetMeter execute targets after: 
17:08:47.963287 [trace] effects : Effect.cpp:484 :  ... OBJ_PLANET 3150: Gienah β I  ... METER_SHIELD: Cur: 1 Init: 0  ... METER_TROOPS: Cur: 39.9 Init: 35.9 ... last attacked on turn: 117

Execute SetMeter effect: 
SetTroops value = min(Value + 1.00, Target.MaxTroops))
... SetMeter execute targets before: 
17:08:47.969291 [trace] effects : Effect.cpp:437 :  ... OBJ_PLANET 3150: Gienah β I .... METER_TROOPS: Cur: 39.9 Init: 35.9  ... last attacked on turn: 117
... SetMeter execute targets after: 
17:08:47.971794 [trace] effects : Effect.cpp:484 :  ... OBJ_PLANET 3150: Gienah β I  ... METER_TROOPS: Cur: 40.9 Init: 35.9  ... last attacked on turn: 117
```

but still, the StoreTargetCauses stage is recognizing that Gienah β I qualifies for the portions of DEF_GARRISON_2 that do not have the LastTurnBattleHere check, but does not qualify for the portions that *do* have that check.

```
17:08:41.432064 [trace] effects : Universe.cpp:1152 : StoreTargetsAndCausesOfEffectsGroups: effects_group:   specific_cause: DEF_GARRISON_2  sources: Glorious Sceptrum I (9086)  )
Generated new target set, for Condition:  that is a Planet and that belongs to source's owner empire and that has Population of at least 0.010000    targets: (Minkar γ I (868)  Minkar γ II (882)  Minkar γ Asteroids (896)  Gienah β I (3150)  Bacabs α III (6272)  Tyr I (8358)  Bacabs α V (6300)  Bacabs α I (6314)  Marfik α I (7980)  Phaet α I (6384)  Aten β I (7196)  Glorious Sceptrum I (9086)  )

17:08:42.874742 [trace] effects : Universe.cpp:1152 : StoreTargetsAndCausesOfEffectsGroups: effects_group:   specific_cause: DEF_GARRISON_2  sources: Glorious Sceptrum I (9086)  )
Generated new target set, for Condition:  that is a Planet and that belongs to source's owner empire and when local condition candidate's system's last turn a battle occurred here < the current turn    targets: (Minkar γ I (868)  Minkar γ II (882)  Minkar γ Asteroids (896)  Gienah β I (3150)  Bacabs α III (6272)  Bacabs α II (6286)  Bacabs α V (6300)  Bacabs α I (6314)  Bacabs α IV (6328)  Phaet α I (6384)  Aten β I (7196)  Glorious Sceptrum I (9086)  Marfik α I (7980)  Tyr I (8358)  )
```

So it seems like somehow the TargetSet (objects matching scope) is wrongly getting modified between being initially stored and when the effects are actually executed.


I also double checked the condition of the planet before combat, and already at that point the Shields were at Cur: 1, Init: 0, which seems wrong because there had also been combat on the turn prior, so the shields should have been Cur:0 Init 0 the entire time it seems to me (they did get knocked down to Cur 0 during this turn's combat).   So there must be a surviving bit of C++ code that bumped up the shields by 1 point, it seems to me. The troops however during the combat stage still had Cur and Init values the same.

**Comment by dbenage-cx** [comment]
wrt troops, I do not suspect an errant meter gain in c++ since if all 5 effects are removed/`NoOp`ed then troops do not increase.

**Comment by Dilvish-fo** [comment]
> wrt troops, I do not suspect an errant meter gain in c++ since if all 5 effects are removed/NoOped then troops do not increase.

And you only removed the portions that do regen, not the portions that set the respective Max meter?  I don't understand how we could have shields Init: 0 Cur:1 prior to combat (prior to any meter Effects being executed that turn) if not for something happening in c++

**Comment by dbenage-cx** [comment]
yes, here is [diff](https://gist.github.com/dbenage-cx/a8a92e961309ca7507f5de5a384c3816)
Using a previous game, after the first next turn, no regen for troops.
To me this implies that Current is incorrect (edit: I mean that effects are incorrectly applied to Current)

**Comment by Dilvish-fo** [comment]
nm, I just repeated that check myself to be sure, taking out only the regen portions of the scripts, and indeed it did prevent the shield meter from having init 0 cur 1 before combat.

**Comment by Dilvish-fo** [comment]
That first saved game was rather unwieldy for debugging with effects trace logging on; here is a much smaller and more manageable one.

[__troop_regen_check2.zip](https://github.com/freeorion/freeorion/files/1746704/__troop_regen_check2.zip)

**Comment by Dilvish-fo** [comment]
Also worth noting, I tried running with single-threaded effects processing (everything single threaded actually) but still got the same regens

**Comment by dbenage-cx** [comment]
@Dilvish-fo  Any issue with [checking vs previous turn](https://gist.github.com/dbenage-cx/0e4c3f4057c5b721e8d3bc572d7facfc)?

**Comment by Dilvish-fo** [comment]
> Any issue with checking vs previous turn?

That does stop the regen during combat, but also stops the regen for an extra turn after the combat ends, which is the problem we had originally noticed and why we had dropped the -1.  So the patch is not a true solution, and the way the scripts are written now is the way that ought to work.  But the results from the patch are certainly another good clue to help narrow down what is going wrong, thanks for suggesting to check that.

I suppose the regen-surpressed-for-an-extra-turn would distort gameplay less while we continue trying to figure out why the scripts are not working right, but I'm doubtful that simply reduced distortion should be our priority.   As far as debugging and testing goes, I think that the manner the problem manifests with the scripts in their current form is much easier to arrange and observe, and therefor hopefully debug.  

But anyone looking to simply enjoy a game would probably be best off applying your diff.  (Note to any such folks: follow the link, use the 'Raw' button towards the right to save a copy to your FreeOrion directory, and then from the command line in that directory run ```git apply x.diff```)

**Comment by dbenage-cx** [comment]
From addition in c2afb236:
```diff
diff --git a/server/ServerApp.cpp b/server/ServerApp.cpp
index fcf72884c..2fb6064c9 100644
--- a/server/ServerApp.cpp
+++ b/server/ServerApp.cpp
@@ -3266,7 +3266,7 @@ void ServerApp::PostCombatProcessTurns() {
     TraceLogger(effects) << objects.Dump();
 
     // redo meter estimates to hopefully be consistent with what happens in clients
-    m_universe.UpdateMeterEstimates(false);
+    // m_universe.UpdateMeterEstimates(false);
 
     TraceLogger(effects) << "ServerApp::PostCombatProcessTurns After Final Meter Estimate Update: ";
     TraceLogger(effects) << objects.Dump();
```
Possibly breaks something else, though not noticing any other differences yet.

**Comment by Dilvish-fo** [comment]
Ah, yes, that is exactly it, whew!  Good to get that figured out.  Presumably taking it out might in some cases result in the target value discrepancies Geoff was trying to stop with it, but that is a far far lesser issue than just breaking various Effects.

Assuming this does lead to some discrepancies coming back, let's brainstorm other possible approaches to stopping it.  To my knowledge, there are two sources of discrepancies.  One is from the server having knowledge of things in the Universe that the client empire is unaware of, and I doubt this change (or any change) can or even should do away with those discrepancies.  

The other was from the fact that population growth is handled after the last meter update-- we have combat, apply all effects (including meters) then resolve supply & blockades, then apply only meter effects to update meters subject to supply constraints, then we do pop growth so that they can have the proper target pop meter.   One possible solution is just on the client side-- with a bit of work I think it could figure out what portion of planetary meter discrepancy is probably due to pop growth (it could temporarily reset pop to previous value, check meter estimates, then reset pop to current value) and then label that portion (which might be all of the discrepancy) as due to pop growth contribution.  That might result in a noticeable amount of extra UI work/lag, and so perhaps it should be an Option about whether to do that.

Another possibility might be to just move this ```m_universe.UpdateMeterEstimates(false);``` to be just after pop growth but still on the same turn, rather than after advancing the turn (which caused all these troubles).

**Comment by Dilvish-fo** [comment]
> Another possibility might be to just move this m_universe.UpdateMeterEstimates(false); to be just after pop growth but still on the same turn, rather than after advancing the turn (which caused all these troubles).

Yes, indeed, that seems to work like a charm for me.  At least, the regen acts correctly, and presumably it would still be eliminating or at least reducing the meter discrepancies it was aimed at.


```diff
diff --git a/server/ServerApp.cpp b/server/ServerApp.cpp
index fcf7288..05b0857 100644
--- a/server/ServerApp.cpp
+++ b/server/ServerApp.cpp
@@ -3214,6 +3214,12 @@ void ServerApp::PostCombatProcessTurns() {
     TraceLogger(effects) << "!!!!!!!!!!!!!!!!!!!!!!AFTER GROWTH AND CLAMPING";
     TraceLogger(effects) << objects.Dump();

+    // redo meter estimates to hopefully be consistent with what happens in clients
+    m_universe.UpdateMeterEstimates(false);
+
+    TraceLogger(effects) << "ServerApp::PostCombatProcessTurns After Final Meter Estimate Update: ";
+    TraceLogger(effects) << objects.Dump();
+
     // store initial values of meters for this turn.
     m_universe.BackPropagateObjectMeters();
     empires.BackPropagateMeters();
@@ -3262,16 +3268,6 @@ void ServerApp::PostCombatProcessTurns() {
     m_universe.UpdateEmpireObjectVisibilities();


-    TraceLogger(effects) << "ServerApp::PostCombatProcessTurns Before Final Meter Estimate Update: ";
-    TraceLogger(effects) << objects.Dump();
-
-    // redo meter estimates to hopefully be consistent with what happens in clients
-    m_universe.UpdateMeterEstimates(false);
-
-    TraceLogger(effects) << "ServerApp::PostCombatProcessTurns After Final Meter Estimate Update: ";
-    TraceLogger(effects) << objects.Dump();
-
-
     // copy latest visible gamestate to each empire's known object state
     m_universe.UpdateEmpireLatestKnownObjectsAndVisibilityTurns();
```

**Comment by Dilvish-fo** [comment]
Since it doesn't seem to be just meter "estimates" that get updated, this change would probably result in a small bit of substantive meter change from where things would be without this extra call.  Some wouldn't really matter, like the industry meter which would get updated after combat before its values are used anyways, but some would matter-- at least, this could probably result in a small bit more increase to a troops meter than would happen without it.  But it's not like that would be a "surprise" difference to anyone.

**Comment by Dilvish-fo** [comment]
I suppose that to avoid the ~~problem~~ difference that I noted just above, we could instead move this update call to be just *after* the back-propagation of meters (but still before turn advance).

**Comment by Dilvish-fo** [comment]
I just noticed a particular variation of population-dependent discrepancy which is very common-- when a colony building first completes, on that first turn its target meters are still at zero, which leads the client to report a big negative 'Unknown' discrepancy.  That could probably be dealt with as I suggest above, instead of just calling it an 'Unknown' discrepancy, having the client label it "Transient Population Effects" or somesuch.  

We might possibly also be able to fix this particular one by adjusting various effectsgroup priorities (but my first experiment on that, to simply make the LifecycleManip portion of colony buildings have an early priority, did not work out right).

I think we should put in a PR with the changes above to wrap up this regen Issue, and either reopen Issue #1894 or open a new Issue more generally covering these meter discrepancy issues.

**Comment by Dilvish-fo** [comment]
Now that we know the cause, I think we should make sure we get this fixed before the next weekly build. 
 @dbenage-cx you're the one who figured out what was actually causing the regen, so I'll suggest you submit a PR.

**Comment by dbenage-cx** [comment]
Posted a form of your first iteration as WIP, I did not notice the later issues you describe when testing so going through more testing.  Some weirdness with map resource values (which was hard to define) seems to have resided.

**Comment by geoffthemedio** [comment]
Having the meter updates after the turn increment is the correct place to do them. The fix for this should be to figure out why the current meter values are being used during the pre-combat turn processing instead of the initial meter values, or perhaps why meters are being back propagated, which converts the current into intial meter values.

Possibly during ground combat processing, the back propagation should be more restricted to planets that actually had combat, not just those with any troops on them, and only apply to the troop and rebel troop meters, not all meters.

https://github.com/freeorion/freeorion/blob/master/server/ServerApp.cpp#L2666

Edit: More likely: https://github.com/freeorion/freeorion/blob/master/server/ServerApp.cpp#L2034

Edit2: trying adding a meter reset at the start of pre-combat turn processing...

**Comment by Dilvish-fo** [comment]
Perhaps I'm forgetting something, but I'm not sure what you are referring to by " why the current meter values are being used during the pre-combat turn processing instead of the initial meter values"
I think that extra backpropagation you point out looks like a very likely solution.  And I agree about the additional restrictions you suggest. 

 If no one else gets to this by later today I'll try to write up a PR.  Reminder for myself or whoever does that--  do a search to double check for any additional new backpropagations.

**Comment by geoffthemedio** [comment]
I unintentionally pushed to master https://github.com/freeorion/freeorion/commit/e97989efa32dfc0afb1acfd44980c7f573c33f6e just now (was supposed to be a separate branch, but the git client did something different after a recent update). Seems to help. Resetting all meters caused problems because there were no combats occurring at all because everything had 0 detection. This just resets active paired meters, essentially to the state they were before the predicted values were generated. Adjusting when / if back propagation occurs during combat or other processing might work better, though I'm not sure.

**Comment by Dilvish-fo** [comment]
https://github.com/freeorion/freeorion/commit/e97989efa32dfc0afb1acfd44980c7f573c33f6e does indeed seem to me to have fixed the problem, and it makes sense to me that it would.   I don't see any reason to be particularly concerned about the remaining post-ground-combat back propagation now, although from a reliability it might still be best for it to be limited to the troop meters.  

@dbenage-cx do you concur this seems fixed now?
